File: bench/scan_delims.lua
-- BOT-ASSISTED

local ffi = require("ffi")
local CAP = 2 ^ 16
local RUNS = (2 ^ 27) / (CAP / 16)
local WANT = (2 ^ 27) * 9
local pat = "{a,b:c[d]e} \n\tfg"
local buf = ffi.new("uint8_t[?]", CAP)
local dels = ffi.new("uint8_t[256]")
local LBRACE = string.byte("{")
local RBRACE = string.byte("}")
local LBRACK = string.byte("[")
local RBRACK = string.byte("]")
local COLON = string.byte(":")
local COMMA = string.byte(",")
local SPACE = string.byte(" ")
local LF = string.byte("\n")
local TAB = string.byte("\t")

dels[LBRACE] = 1
dels[RBRACE] = 1
dels[LBRACK] = 1
dels[RBRACK] = 1
dels[COLON] = 1
dels[COMMA] = 1
dels[SPACE] = 1
dels[LF] = 1
dels[TAB] = 1

for ind = 0, CAP - 1 do buf[ind] = pat:byte((ind % #pat) + 1) end

local function scan(buf, cap, dels)
  local out = 0
  for ind = 0, cap - 1 do
    out = out + dels[buf[ind]]
  end
  return out
end

local out = 0
for _ = 1, RUNS do out = out + scan(buf, CAP, dels) end
if out ~= WANT then error("mismatch: expected " .. WANT .. "; got " .. out) end

File: bench/scan_delims.mjs
// BOT-ASSISTED

const CAP = 1 << 16
const RUNS = (1 << 27) / (CAP / 16)
const WANT = (1 << 27) * 9
const pat = Buffer.from("{a,b:c[d]e} \n\tfg")
const buf = new Uint8Array(CAP)
const dels = new Uint8Array(256)
const LBRACE = "{".charCodeAt(0)
const RBRACE = "}".charCodeAt(0)
const LBRACK = "[".charCodeAt(0)
const RBRACK = "]".charCodeAt(0)
const COLON = ":".charCodeAt(0)
const COMMA = ",".charCodeAt(0)
const SPACE = " ".charCodeAt(0)
const LF = "\n".charCodeAt(0)
const TAB = "\t".charCodeAt(0)

dels[LBRACE] = 1
dels[RBRACE] = 1
dels[LBRACK] = 1
dels[RBRACK] = 1
dels[COLON] = 1
dels[COMMA] = 1
dels[SPACE] = 1
dels[LF] = 1
dels[TAB] = 1

for (let ind = 0; ind < CAP; ind++) buf[ind] = pat[ind % pat.length]

function scan(buf, dels) {
  let out = 0
  for (let ind = 0; ind < buf.length; ind++) out += dels[buf[ind]]
  return out
}

let out = 0
for (let run = 0; run < RUNS; run++) out += scan(buf, dels)
if (out !== WANT) throw Error(`mismatch: expected ${WANT}; got ${out}`)
